### PR TITLE
CLI draft and sync bug fixes

### PR DESCRIPTION
**Pull request by thezhe**
Updates: 
1. fixed issue where random errors would pop up due to non-null terminated char*
2. refactored sync() NOTE: I did not change test_cpisync3node.cc, so the correct usage is in niok.cc.
3. added CLI (niok.cc)
4. moved calls to commit() outside of put, get, remove so now you should call processLogEntry() after each set of operations to commit them. This is so that the int EOC gets updated properly after each commit.
Issues:
1. Could not get sync() to run on a another thread while I parsed input. It seems as if CPISync doesn't like it when it is preempted and must run all the way through. As you can see in the attached picture, I had to manually tell one of the nodes to sync for each operation.
CLI

### CODE REVIEW

File: src/CPISync/GossipNode.cc [hunk 1]
```diff
@@ -5,31 +5,26 @@ namespace niok
 namespace cpisync
 {
 
-GossipNode::GossipNode(string nodeName, int spaceId, string path,
-               hash<string> hashFunc, vector<string> initialElems)
+GossipNode::GossipNode(string nodeName, int spaceId, string path, //db
+                       hash<string> hashFunc, int numCharHash, int numCharEntry, //hashing
+                       vector<string> initialEntries) //initial log entries
+    :name_ (nodeName), rootPath_ (path), NUM_CHAR_HASH(numCharHash),NUM_CHAR_ENTRY(numCharEntry) //const variables
 {
```
Comment (wfystx) [inline src/CPISync/GossipNode.cc:12]: good usage

File: src/CPISync/GossipNode.cc [hunk 2]
```diff
@@ -79,83 +74,67 @@ bool GossipNode::put(std::string key, std::string value)
     string logEntry = keyValueToLog(key, value, "P");
     log_.push_back(logEntry);
     hashDefs[to_string(strHash(logEntry))] = logEntry;
-    
-    return commit(logEntry);
+    return true;
```
Comment (wfystx) [inline src/CPISync/GossipNode.cc:77]: when to return false? if always true. we can just make it void.

File: src/CPISync/GossipNode.cc [hunk 3]
```diff
@@ -79,83 +74,67 @@ bool GossipNode::put(std::string key, std::string value)
...
     log_.push_back(logEntry);
     hashDefs[to_string(strHash(logEntry))] = logEntry;
-    
-    return commit(logEntry);
-}
-
-void addNeighbor(IPv4 *ip) 
-{
-    neighbors_.insert(ip);
+    return true;
```
Comment (wfystx) [inline src/CPISync/GossipNode.cc:85]: ditto

File: src/CPISync/GossipNode.cc [hunk 4]
```diff
@@ -79,83 +74,67 @@ bool GossipNode::put(std::string key, std::string value)
...
 }
 
-void GossipNode::sync(string host, int NUM_CHAR, bool server)
+void GossipNode::sync(string host, bool server, int times)
```
Comment (wfystx) [inline src/CPISync/GossipNode.cc:88]: when is 'times' being used?

File: src/CPISync/GossipNode.cc [hunk 5]
```diff
@@ -179,26 +160,23 @@ vector<string> GossipNode::logToKeyValue(string log) {
...
+        if (!commit(log_[i]))
         {
             cout << "this LogEntry commited failed : ";
-            cout << logEntry << endl;
+            cout << log_[i] << endl;
         }
     }
-    EOC = i;
+    EOC = logSize;
 }
```
Comment (wfystx) [inline src/CPISync/GossipNode.cc:173]: good refactoring

File: src/CPISync/GossipNode.cc [hunk 6]
```diff
@@ -79,83 +74,67 @@ bool GossipNode::put(std::string key, std::string value)
...
+        ++i;
     }
-    EOL = log_.size();
-    // cout << endl;
+    //update EOL
+    EOL = logSize;
 }
 
-string GossipNode::keyValueToLog(string key, string value, string op) {
+string GossipNode::keyValueToLog(string key, string value, string op)
```
Comment (zhou-1) [inline src/CPISync/GossipNode.cc:133]: So currently it is still a string, we should convert it to Log.

File: src/CPISync/GossipNode.cc [hunk 7]
```diff
@@ -179,26 +160,23 @@ vector<string> GossipNode::logToKeyValue(string log) {
 
 void GossipNode::processLogEntry()
 {
-    int i = EOC;
-    for (i; i < log_.size(); i++) {
-        string logEntry = log_[i];
-        if (!commit(logEntry)) 
+    const int logSize = log_.size();
```
Comment (zhou-1) [inline src/CPISync/GossipNode.cc:163]: Good one using const here.

**Pull request merged**